jung-visualization/src/main/java/edu/uci/ics/jung/visualization/subLayout/TreeCollapser.java

```diff
@@ -49,6 +52,29 @@ public static void expand(MutableCTreeNetwork tree, CTreeNetwork subTree) {
             ? tree.inEdges(subTree).iterator().next() // THERE CAN BE ONLY ONE
             : null;
     tree.removeNode(subTree);
-    TreeUtils.addSubTree(tree, subTree, parent.orElse(null), parentEdge);
+    if (parent.isPresent()) {
+      TreeUtils.addSubTree(tree, subTree, parent.get(), parentEdge);
+    } else {
+      // tree is empty, so just copy all of subTree into tree
```

Comment:
A couple of notes here:

(1) The API feels a bit inconsistent: collapse() returns a tree, but expand() does not.  We should consider having expand() return a tree.
(2) If `tree` is empty, it seems like we should just be returning subtree rather than copying it into tree.

Thoughts?

Reply:
@jrtom The commit where I'm having problems is https://github.com/jrtom/jung/pull/204/commits/04bca188381cfd2cfc77b835642cecedf376d94e.

The problem I'm having is as follows.

If I compiile and run `TreeCollapseDemo` from that commit, and if one selects the root node and clicks "Collapse", then the tree collapses fully and everything seems fine.

But if one then attempts to expand the tree back to how before with the "Expand" button, then the following exception is thrown.

Do you have any suggestions as to how I may be able to proceed, or any other thoughts?

```
Exception in thread "AWT-EventQueue-0" java.lang.IllegalArgumentException
	at com.google.common.base.Preconditions.checkArgument(Preconditions.java:120)
	at edu.uci.ics.jung.layout.algorithms.TreeLayoutAlgorithm.buildTree(TreeLayoutAlgorithm.java:92)
	at edu.uci.ics.jung.layout.algorithms.TreeLayoutAlgorithm.visit(TreeLayoutAlgorithm.java:84)
	at edu.uci.ics.jung.layout.model.AbstractLayoutModel.accept(AbstractLayoutModel.java:89)
	at edu.uci.ics.jung.visualization.BaseVisualizationModel.setNetwork(BaseVisualizationModel.java:169)
	at edu.uci.ics.jung.samples.TreeCollapseDemo.lambda$new$3(TreeCollapseDemo.java:167)
	at javax.swing.AbstractButton.fireActionPerformed(AbstractButton.java:2022)
	at javax.swing.AbstractButton$Handler.actionPerformed(AbstractButton.java:2348)
	at javax.swing.DefaultButtonModel.fireActionPerformed(DefaultButtonModel.java:402)
	at javax.swing.DefaultButtonModel.setPressed(DefaultButtonModel.java:259)
	at javax.swing.plaf.basic.BasicButtonListener.mouseReleased(BasicButtonListener.java:252)
	at java.awt.Component.processMouseEvent(Component.java:6533)
	at javax.swing.JComponent.processMouseEvent(JComponent.java:3324)
	at java.awt.Component.processEvent(Component.java:6298)
	at java.awt.Container.processEvent(Container.java:2237)
	at java.awt.Component.dispatchEventImpl(Component.java:4889)
	at java.awt.Container.dispatchEventImpl(Container.java:2295)
	at java.awt.Component.dispatchEvent(Component.java:4711)
	at java.awt.LightweightDispatcher.retargetMouseEvent(Container.java:4889)
	at java.awt.LightweightDispatcher.processMouseEvent(Container.java:4526)
	at java.awt.LightweightDispatcher.dispatchEvent(Container.java:4467)
	at java.awt.Container.dispatchEventImpl(Container.java:2281)
	at java.awt.Window.dispatchEventImpl(Window.java:2746)
	at java.awt.Component.dispatchEvent(Component.java:4711)
	at java.awt.EventQueue.dispatchEventImpl(EventQueue.java:758)
	at java.awt.EventQueue.access$500(EventQueue.java:97)
	at java.awt.EventQueue$3.run(EventQueue.java:709)
	at java.awt.EventQueue$3.run(EventQueue.java:703)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:80)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:90)
	at java.awt.EventQueue$4.run(EventQueue.java:731)
	at java.awt.EventQueue$4.run(EventQueue.java:729)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:80)
	at java.awt.EventQueue.dispatchEvent(EventQueue.java:728)
	at java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:201)
	at java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:116)
	at java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:105)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:93)
	at java.awt.EventDispatchThread.run(EventDispatchThread.java:82)
```

Reply:
I'm also having trouble understanding `TreeCollapser.collapse`. It seems to change its first parameter `tree` (which I think is where the collapsing is happening), but then it proceeds to return an internal variable `subTree` rather than the changed `tree`, so I'm rather lost.

Reply:
Yeah, the return value on `collapse()` is weird; I'm not sure why it's returning `subTree` either (and there's no documentation to clarify it).  It's not uncommon to have a parameter that you modify and also return (to provide for a more fluent API, generally), but if you're going to return some other thing you ought to specify what, and why.

I'll have to take a closer look at the exception above in order to get a handle on what's going on; I'll respond again once I've had a chance to dig into it.  Thanks for bringing this to my attention.

Reply:
Finally got a few minutes to look at this. tl;dr:  I think I've figured it out, but please follow my reasoning and see if I've missed something.

The exception is thrown in `TreeLayoutAlgorithm.buildTree()` if the supplied graph has no roots.  This code is reached via `TreeLayoutAlgorithm.visit()`, which itself gets triggered by updating the graph in `BaseVisualizationModel.setNetwork()`.

This can happen either because (a) the graph is empty, (b) the graph has no nodes with no predecessor (i.e., roots), or conceivably (c) there is a bug in `TreeUtils.roots()`.

The relevant changes in `TreeCollapseDemo` are from this:

    TreeCollapser.expand(graph, (CTreeNetwork) v);
    vv.getModel().setNetwork(graph);

to this:

    CTreeNetwork expandedTree = TreeCollapser.expand(graph, (CTreeNetwork) v);
    LayoutModel<Object> objectLayoutModel = vv.getModel().getLayoutModel();
    objectLayoutModel.set(expandedTree, objectLayoutModel.apply(v));
    vv.getModel().setNetwork(graph, true);

There are a couple of potential problems here.

(1) The code is now explicitly updating the position of the expanded tree in `LayoutModel` to that of `v` in that model.  I'm not sure why that is the case.
(Note that the `LayoutModel`'s graph is updated in the `setNetwork()` call, which might 
invalidate the change you made there, not 100% sure at the moment.)
(2) The parameter being passed to `setNetwork` is `graph`; you probably wanted `expandedTree`.

My guess is that while (1) might not be needed and might be causing its own problems if they weren't being overshadowed, (2) is probably the source of the problem you reported.

LMK if this makes sense.